plugins/filter_aws/aws.c

```diff
@@ -182,191 +206,279 @@ static int cb_aws_init(struct flb_filter_instance *f_ins,
     return 0;
 }
 
-/* Get an IMDSv2 token */
-static int get_ec2_token(struct flb_filter_aws *ctx)
+/* Get VPC metadata, it called IMDS twice.
+ * First is for getting the Mac ID and combine into the path for VPC.
+ * Second call is using the VPC path to get the VPC id
+ */
+static int get_vpc_metadata(struct flb_filter_aws *ctx)
 {
     int ret;
-    size_t b_sent;
-    struct flb_connection *u_conn;
-    struct flb_http_client *client;
+    flb_sds_t mac_id = NULL;
+    flb_sds_t tmp;
+    flb_sds_t vpc_path;
 
-    u_conn = flb_upstream_conn_get(ctx->ec2_upstream);
-    if (!u_conn) {
-        flb_plg_error(ctx->ins, "connection initialization error");
+    mac_id = flb_aws_imds_get_vpc_id(ctx->client_imds);
+    if (!mac_id) {
+        flb_plg_error(ctx->ins, "Could not retrieve vpc id from IMDS");
         return -1;
     }
 
-    /* Compose HTTP Client request */
-    client = flb_http_client(u_conn, FLB_HTTP_PUT,
-                             FLB_FILTER_AWS_IMDS_V2_TOKEN_PATH,
-                             NULL, 0, FLB_FILTER_AWS_IMDS_HOST,
-                             80, NULL, 0);
-
-    if (!client) {
-        flb_plg_error(ctx->ins, "count not create http client");
-        flb_upstream_conn_release(u_conn);
+    /* the VPC full path should be like:
+     *latest/meta-data/network/interfaces/macs/{mac_id}/vpc-id/"
+     */
+    vpc_path = flb_sds_create_size(70);
+    if (!vpc_path) {
+        flb_errno();
```

Comment:
This change has broken the vpc_id feature. I think you can just use `flb_aws_imds_get_vpc_id`  to get the VPC ID. You don't need to then construct the path and make multiple requests.

That function already does that: https://github.com/fluent/fluent-bit/blob/master/src/aws/flb_aws_imds.c#L219